File: javascript/weather.js
/**
 * World Weather Online — Weather Fetcher (Node.js)
 * =================================================
 * Fetches and displays current weather + 5-day forecast in the terminal.
 *
 * Requirements:
 *   npm install axios chalk
 *
 * Usage:
 *   node weather.js
 *   node weather.js London
 *   node weather.js "New York" 3
 *
 * Set your API key:
 *   export WWO_API_KEY="your_key_here"
 *
 * Get a free key at:
 *   https://www.worldweatheronline.com/weather-api/
 */

const axios = require("axios");

// ─── CONFIG ────────────────────────────────────────────────────────────────────

const API_KEY = process.env.WWO_API_KEY || "your_api_key_here";
const BASE_URL = "https://api.worldweatheronline.com/premium/v1/weather.ashx";

// Try to load chalk for coloured output (optional)
let chalk;
try {
  chalk = require("chalk");
} catch {
  // If chalk isn't installed, use plain text
  chalk = {
    blue: (s) => s, bold: { blue: (s) => s, yellow: (s) => s, cyan: (s) => s },
    cyan: (s) => s, red: (s) => s, green: (s) => s, yellow: (s) => s, dim: (s) => s,
  };
}

// Weather condition → emoji
const ICONS = {
  sunny: "☀️", clear: "🌙", "partly cloudy": "⛅", cloudy: "☁️",
  overcast: "☁️", mist: "🌫️", fog: "🌫️", rain: "🌧️", drizzle: "🌦️",
  snow: "❄️", sleet: "🌨️", thunder: "⛈️", blizzard: "🌨️",
};

function getIcon(description) {
  const desc = description.toLowerCase();
  for (const [key, icon] of Object.entries(ICONS)) {
    if (desc.includes(key)) return icon;
  }
  return "🌡️";
}


// ─── API CALL ──────────────────────────────────────────────────────────────────

async function getWeather(location, days = 5) {
  if (API_KEY === "your_api_key_here") {
    console.error("❌  Please set your API key!");
    console.error("    export WWO_API_KEY='your_key_here'");
    console.error("    Get a free key: https://www.worldweatheronline.com/weather-api/");
    process.exit(1);
  }

  try {
    const response = await axios.get(BASE_URL, {
      params: {
        key: API_KEY,
        q: location,
        format: "json",
        num_of_days: days,
        tp: 24,
        includelocation: "yes",
        cc: "yes",
      },
      timeout: 10000,
    });

    const data = response.data.data;

    // Check for API-level errors
    if (data.error) {
      console.error(`❌  API Error: ${data.error[0].msg}`);
      process.exit(1);
    }

    return data;

  } catch (err) {
    if (err.code === "ENOTFOUND") {
      console.error("❌  No internet connection.");
    } else if (err.code === "ECONNABORTED") {
      console.error("❌  Request timed out.");
    } else if (err.response) {
      console.error(`❌  HTTP ${err.response.status}: ${err.response.statusText}`);
    } else {
      console.error(`❌  Error: ${err.message}`);
    }
    process.exit(1);
  }
}


// ─── DISPLAY ───────────────────────────────────────────────────────────────────

function displayCurrent(current, locationName) {
  const desc = current.weatherDesc[0].value;
  const icon = getIcon(desc);

  console.log("\n" + "─".repeat(50));
  console.log(chalk.bold.yellow(`📍 ${locationName} — Right Now`));
  console.log("─".repeat(50));
  console.log(`${icon}  ${desc}`);
  console.log(`🌡️  Temperature : ${chalk.cyan(current.temp_C + "°C")} / ${current.temp_F}°F  (Feels like ${current.FeelsLikeC}°C)`);
  console.log(`💧  Humidity    : ${current.humidity}%`);
  console.log(`💨  Wind        : ${current.windspeedMiles} mph ${current.winddir16Point}`);
  console.log(`👁️  Visibility  : ${current.visibility} km`);
  console.log(`☀️  UV Index    : ${current.uvIndex}`);
  console.log("─".repeat(50));
}

function displayForecast(weatherDays) {
  console.log(chalk.bold.blue("\n📅 Forecast\n"));

  const header = [
    "Date".padEnd(14),
    "Conditions".padEnd(25),
    "High".padStart(7),
    "Low".padStart(7),
    "Rain%".padStart(7),
    "Wind".padStart(7),
  ].join("  ");

  console.log(chalk.cyan(header));
  console.log("─".repeat(75));

  for (const day of weatherDays) {
    const dateObj = new Date(day.date);
    const dateFmt = dateObj.toLocaleDateString("en-GB", {
      weekday: "short", day: "2-digit", month: "short",
    });

    const desc = day.hourly[0].weatherDesc[0].value;
    const icon = getIcon(desc);
    const rainChance = day.hourly[0].chanceofrain || "N/A";
    const wind = day.hourly[0].windspeedMiles;

    const row = [
      dateFmt.padEnd(14),
      `${icon} ${desc}`.padEnd(25),
      chalk.red((day.maxtempC + "°C").padStart(7)),
      chalk.blue((day.mintempC + "°C").padStart(7)),
      (rainChance + "%").padStart(7),
      (wind + " mph").padStart(7),
    ].join("  ");

    console.log(row);
  }

  console.log("─".repeat(75));
}


// ─── MAIN ──────────────────────────────────────────────────────────────────────

async function main() {
  const location = process.argv[2] || "London";
  const days = parseInt(process.argv[3]) || 5;

  console.log(chalk.bold.blue(`\n🌍 World Weather Online`) + ` — fetching weather for ${chalk.bold(location)}...`);

  const data = await getWeather(location, days);

  // Get location name
  let locationName = location;
  try {
    const area = data.nearest_area[0].areaName[0].value;
    const country = data.nearest_area[0].country[0].value;
    locationName = `${area}, ${country}`;
  } catch {}

  displayCurrent(data.current_condition[0], locationName);
  displayForecast(data.weather);

  console.log(chalk.dim("\nData by World Weather Online — https://www.worldweatheronline.com\n"));
}

main();

File: swift/weather.swift
/**
 * World Weather Online — Weather Dashboard (Swift)
 * ==================================================
 * Fetches and displays current weather + 5-day forecast.
 *
 * Requirements:
 *   Swift 5.5+  (uses async/await and URLSession — no packages needed)
 *   Works on macOS, Linux (Swift for Linux), and as a Swift Package
 *
 * Run (macOS/Linux):
 *   swift weather.swift
 *   swift weather.swift London
 *   swift weather.swift "New York" 3
 *
 * Set your API key:
 *   export WWO_API_KEY="your_key_here"
 *
 * Get a free key at:
 *   https://www.worldweatheronline.com/weather-api/
 */

import Foundation

// ─── CONFIG ───────────────────────────────────────────────────────────────────

let apiKey  = ProcessInfo.processInfo.environment["WWO_API_KEY"] ?? "your_api_key_here"
let baseURL = "https://api.worldweatheronline.com/premium/v1/weather.ashx"

let icons: [String: String] = [
    "sunny":         "☀️",
    "clear":         "🌙",
    "partly cloudy": "⛅",
    "cloudy":        "☁️",
    "overcast":      "☁️",
    "mist":          "🌫️",
    "fog":           "🌫️",
    "rain":          "🌧️",
    "drizzle":       "🌦️",
    "snow":          "❄️",
    "sleet":         "🌨️",
    "thunder":       "⛈️",
    "blizzard":      "🌨️",
]

func getIcon(_ description: String) -> String {
    let desc = description.lowercased()
    for (key, icon) in icons {
        if desc.contains(key) { return icon }
    }
    return "🌡️"
}


// ─── API STRUCTS ──────────────────────────────────────────────────────────────

struct WeatherResponse: Codable {
    let data: WeatherData
}

struct WeatherData: Codable {
    let current_condition: [CurrentCondition]
    let weather: [DayForecast]
    let nearest_area: [NearestArea]?
    let error: [APIError]?
}

struct CurrentCondition: Codable {
    let temp_C: String
    let temp_F: String
    let FeelsLikeC: String
    let humidity: String
    let windspeedMiles: String
    let winddir16Point: String
    let uvIndex: String
    let visibility: String
    let weatherDesc: [Description]
}

struct DayForecast: Codable {
    let date: String
    let maxtempC: String
    let mintempC: String
    let hourly: [HourlyData]
}

struct HourlyData: Codable {
    let weatherDesc: [Description]
    let chanceofrain: String?
    let windspeedMiles: String
}

struct NearestArea: Codable {
    let areaName: [Description]
    let country: [Description]
}

struct Description: Codable {
    let value: String
}

struct APIError: Codable {
    let msg: String
}


// ─── API CALL ─────────────────────────────────────────────────────────────────

func fetchWeather(location: String, days: Int) async throws -> WeatherData {
    if apiKey == "your_api_key_here" {
        fputs("❌  Please set your API key!\n", stderr)
        fputs("    export WWO_API_KEY='your_key_here'\n", stderr)
        fputs("    Get a free key: https://www.worldweatheronline.com/weather-api/\n", stderr)
        exit(1)
    }

    var components = URLComponents(string: baseURL)!
    components.queryItems = [
        URLQueryItem(name: "key",             value: apiKey),
        URLQueryItem(name: "q",               value: location),
        URLQueryItem(name: "format",          value: "json"),
        URLQueryItem(name: "num_of_days",     value: "\(days)"),
        URLQueryItem(name: "tp",              value: "24"),
        URLQueryItem(name: "includelocation", value: "yes"),
        URLQueryItem(name: "cc",              value: "yes"),
    ]

    var request = URLRequest(url: components.url!, timeoutInterval: 10)
    request.setValue("WWO-Swift-Client/1.0", forHTTPHeaderField: "User-Agent")

    let (data, response) = try await URLSession.shared.data(for: request)

    guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
        throw URLError(.badServerResponse)
    }

    let result = try JSONDecoder().decode(WeatherResponse.self, from: data)

    if let errors = result.data.error, !errors.isEmpty {
        fputs("❌  API Error: \(errors[0].msg)\n", stderr)
        exit(1)
    }

    return result.data
}


// ─── DISPLAY ──────────────────────────────────────────────────────────────────

func displayCurrent(_ current: CurrentCondition, locationName: String) {
    let desc = current.weatherDesc[0].value
    let icon = getIcon(desc)

    print("\n" + String(repeating: "─", count: 50))
    print("📍 \(locationName) — Right Now")
    print(String(repeating: "─", count: 50))
    print("\(icon)  \(desc)")
    print("🌡️  Temperature : \(current.temp_C)°C / \(current.temp_F)°F (Feels like \(current.FeelsLikeC)°C)")
    print("💧  Humidity    : \(current.humidity)%")
    print("💨  Wind        : \(current.windspeedMiles) mph \(current.winddir16Point)")
    print("👁️  Visibility  : \(current.visibility) km")
    print("☀️  UV Index    : \(current.uvIndex)")
    print(String(repeating: "─", count: 50))
}

func displayForecast(_ weatherDays: [DayForecast]) {
    print("\n📅 Forecast\n")
    print(String(format: "%-14s %-25s %7s %7s %7s", "Date", "Conditions", "High", "Low", "Rain%"))
    print(String(repeating: "─", count: 65))

    let formatter = DateFormatter()
    formatter.dateFormat = "yyyy-MM-dd"
    let displayFormatter = DateFormatter()
    displayFormatter.dateFormat = "EEE dd MMM"

    for day in weatherDays {
        let date    = formatter.date(from: day.date) ?? Date()
        let dateFmt = displayFormatter.string(from: date)
        let desc    = day.hourly[0].weatherDesc[0].value
        let icon    = getIcon(desc)
        let rain    = day.hourly[0].chanceofrain ?? "N/A"

        print(String(format: "%-14s %-25s %7s %7s %7s",
            dateFmt,
            "\(icon) \(desc)",
            "\(day.maxtempC)°C",
            "\(day.mintempC)°C",
            "\(rain)%"
        ))
    }

    print(String(repeating: "─", count: 65))
}


// ─── MAIN ─────────────────────────────────────────────────────────────────────

let location = CommandLine.arguments.count > 1 ? CommandLine.arguments[1] : "London"
let days     = CommandLine.arguments.count > 2 ? Int(CommandLine.arguments[2]) ?? 5 : 5

print("\n🌍 World Weather Online — fetching weather for \(location)...")

// Run async code from synchronous context
let semaphore = DispatchSemaphore(value: 0)

Task {
    do {
        let data = try await fetchWeather(location: location, days: days)

        var locationName = location
        if let area    = data.nearest_area?.first?.areaName.first?.value,
           let country = data.nearest_area?.first?.country.first?.value {
            locationName = "\(area), \(country)"
        }

        displayCurrent(data.current_condition[0], locationName: locationName)
        displayForecast(data.weather)

        print("\nData by World Weather Online — https://www.worldweatheronline.com\n")
    } catch {
        fputs("❌  Error: \(error.localizedDescription)\n", stderr)
        exit(1)
    }
    semaphore.signal()
}

semaphore.wait()
